File: cpp/dynamic_programming/fibonacci_memoization.cpp
#include <iostream>

using namespace std;

int fibonacciMemoization(int num, int dp[]){
    if (num <= 1) {
        return dp[num]=num;
    }
    if(dp[num]!=0){
        return dp[num];
    }
    int left = fibonacciMemoization(num-1, dp);
    int right = fibonacciMemoization(num-2, dp);
    return dp[num]=left+right;

}

int main(){
    int num, res=0;
    cin>>num;
    int dp[num+1] = {0};
    res = fibonacciMemoization(num, dp);
    cout<< res<<endl;
    for(int i=0; i<(sizeof(dp)/sizeof(int)); i++){
        cout << dp[i] <<' ';
    }
}

File: python/dynamic_programming/fibonacci_memoization.py
def fibonacci_memoization(num, arr):
	if num<=1:
		arr[num] = num
		return arr[num]
	elif num in arr:
		return arr[num]
	else:
		left = fibonacci_memoization(num-1, arr)
		right = fibonacci_memoization(num-2, arr)
		arr[num] = left + right
	return arr[num]


num = int(input())
arr = {}
res = fibonacci_memoization(num, arr)
print(res)
print(arr)
